sycl-fusion/jit-compiler/lib/fusion/FusionPipeline.cpp

```diff
@@ -141,3 +142,51 @@ FusionPipeline::runFusionPasses(Module &Mod, SYCLModuleInfo &InputInfo,
 
   return std::make_unique<SYCLModuleInfo>(std::move(*NewModInfo.ModuleInfo));
 }
+
+bool FusionPipeline::runMaterializerPasses(
+    llvm::Module &Mod, std::vector<unsigned char> &SpecConstBlob) {
+  PassBuilder PB;
+  LoopAnalysisManager LAM;
+  FunctionAnalysisManager FAM;
+  CGSCCAnalysisManager CGAM;
+  ModuleAnalysisManager MAM;
+  PB.registerModuleAnalyses(MAM);
+  PB.registerCGSCCAnalyses(CGAM);
+  PB.registerFunctionAnalyses(FAM);
+  PB.registerLoopAnalyses(LAM);
+  PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
+  // Base the pipeline on O3 opt level.
+  ModulePassManager MPM =
+      PB.buildPerModuleDefaultPipeline(OptimizationLevel::O3);
```

Comment:
So this is something I don't quite get and need to investigate further.
My expectation was to be able to set a default level of optimization (`O3`) and only insert a desired pass at the correct stage of the pipelin (in the case at hand that means the materializer pass,  at [the start of the pipeline](https://llvm.org/doxygen/classllvm_1_1PassBuilder.html#a2c79d6634d4333b578eee9468600545f)). 
However that did not yield expected results and I ended up adding passes manually.

Reply:
Maybe this can help? https://llvm.org/docs/NewPassManager.html#inserting-passes-into-default-pipelines

My naive assumption was that what you're currently doing is adding all passes for an `O3` pipeline, then your materializer pass and then another collection of optimization passes. So you would essentially run an `O3` pipeline before spec constant materialization.

Reply:
Lukas is right I think. Rest LGTM.

Reply:
I've removed the forced `O3` pipeline and kept the desired passes in a default one. I'll move the investigation as to why adding the materializer to `registerPipelineStartEPCallback` of `O3` pipeline did not yield expected results into an internal ticket.